modules/dnn/src/layers/nonzero_layer.cpp

```diff
@@ -0,0 +1,281 @@
+            if (rank > 1)
+            {
+                int d = rank - 2;
+                for (; d >= 0; --d)
+                {
+                    coord[d] += 1;
+                    if (coord[d] < dims[d])
+                        break;
+                    coord[d] = 0;
+                }
+            }
+        }
+    });
+}
+
+template <typename T>
+static void emitIndicesStripes(const T* data,
+                               size_t total,
+                               int rank,
+                               const std::vector<int>& dims,
+                               const std::vector<int>& strides,
+                               const std::vector<size_t>& nzstart,
+                               int64* y)
+{
+    emitIndicesStripes<T>(data, total, rank, dims, strides, nzstart, y, [](T v){ return v != (T)0; });
+}
+}
+
+class NonZeroLayerImpl CV_FINAL : public NonZeroLayer
+{
+public:
+    NonZeroLayerImpl(const LayerParams& params) { setParamsFrom(params); }
+
+    virtual bool dynamicOutputShapes() const CV_OVERRIDE
+    {
+        return true;
+    }
+
+    bool supportBackend(int backendId) CV_OVERRIDE
+    {
+        return backendId == DNN_BACKEND_OPENCV;
+    }
+
+    bool getMemoryShapes(const std::vector<MatShape>& inputs, const int /*requiredOutputs*/,
+                         std::vector<MatShape>& outputs, std::vector<MatShape>& /*internals*/) const CV_OVERRIDE
+    {
+        CV_Assert(inputs.size() == 1);
+        const MatShape& in = inputs[0];
+        CV_Assert(!in.empty());
+
+        int rank = (int)in.size();
+        MatShape out = shape(rank, -1);
+        outputs.assign(1, out);
+        return false;
+    }
+
+    void getTypes(const std::vector<MatType>& /*inputs*/, const int requiredOutputs,
+                  const int requiredInternals, std::vector<MatType>& outputs,
+                  std::vector<MatType>& internals) const CV_OVERRIDE
+    {
+        outputs.assign(requiredOutputs, CV_64S);
+    }
+
+    void forward(InputArrayOfArrays in_arr, OutputArrayOfArrays out_arr, OutputArrayOfArrays) CV_OVERRIDE
+    {
+        CV_Assert(in_arr.size().area() == 1);
+        const Mat X = in_arr.getMat(0);
+        CV_Assert(X.data != nullptr);
+
+        const int rank = X.dims;
+        std::vector<int> dims(rank), strides(rank);
+        for (int i = 0; i < rank; ++i) dims[i] = X.size[i];
+
+        if (rank > 0) {
+            strides.back() = 1;
+            for (int i = rank - 2; i >= 0; --i)
+                strides[i] = strides[i + 1] * dims[i + 1];
+        }
+
+        size_t total = X.total();
+
+        const int depth = CV_MAT_DEPTH(X.type());
+        std::vector<size_t> nzcounts;
+        std::vector<size_t> nzstart;
+        size_t nnz = 0;
+        switch (depth)
+        {
+        case CV_8U:  { nnz = computeNonZeroCountsPerStripe<uchar>(X.ptr<uchar>(), total, nzcounts); break; }
+        case CV_8S:  { nnz = computeNonZeroCountsPerStripe<schar>(X.ptr<schar>(), total, nzcounts); break; }
```

Comment:
we don't need separate versions for CV_8U and CV_8S.
we don't need separate versions for CV_16U and CV_16S.
we don't need separate versions for CV_32U and CV_32S.
we don't need separate versions for CV_64U and CV_64S.